src/shims/unix/socket.rs

```diff
@@ -34,6 +37,77 @@ struct Buffer {
     buf_has_writer: bool,
 }
 
+impl EpollTarget for SocketPair {
+    fn check_and_update_readiness(&self, epollin: u32, epollout: u32, epollrdhup: u32) {
+        self.update_readiness(self.check_readiness(epollin, epollout, epollrdhup));
+    }
+    // This function will check the readiness of current file description and return bitmask
+    // that can reflect the readiness.
+    // TODO: passing the flag using function parameter is quite hacky... improve this.
+    fn check_readiness(&self, epollin: u32, epollout: u32, epollrdhup: u32) -> u32 {
+        let readbuf = self.readbuf.borrow();
+        // Start with complement of 0, then unset the flag if we found something is not
+        // doable,
+        let mut readiness: u32 = u32::MAX;
+        // Check if the writeend has closed.
+        if readbuf.buf_has_writer {
+            readiness &= !epollrdhup;
+        }
+
+        // Unset flag if it is unreadable.
+        if readbuf.buf.is_empty() {
+            readiness &= !epollin;
+        }
+
+        // Unset flag if it is not writable.
+        match self.writebuf.upgrade() {
+            Some(writebuf) => {
+                let writebuf = writebuf.borrow();
+                let data_size = writebuf.buf.len();
+                let available_space = MAX_SOCKETPAIR_BUFFER_CAPACITY.strict_sub(data_size);
+                if available_space == 0 {
+                    readiness &= epollout;
+                }
+            }
+            None => {
+                readiness &= epollout;
+            }
+        }
+        readiness
+    }
+    // This function will add epollreturn to ready list if there are matching events in
+    // interest list.
+    fn update_readiness(&self, flag: u32) {
+        for event in &self.epoll_events {
+            if let Some(epoll_event) = event.upgrade() {
+                // TODO: separate the check for each independent flags.
+                if epoll_event.events & flag == flag {
+                    // The file description is self, so the upgrade should always suceed.
```

Comment:
There's no upgrade anymore

Reply:
Hmm, why? ``self.epoll_events`` is ``Vec<Weak<EpollEvent>>``, so I think it needs to be upgraded to gain access.

Reply:
I mean, this comment refers to some code that has no upgrade. We are using the `weak_file_descriptor` without upgrading.